File: solutions/go/proverb/1/proverb.go
// For want of a horseshoe nail, a kingdom was lost, or so the saying goes.
package proverb

import "fmt"

// Proverb given a list of inputs, generate the relevant proverb.
func Proverb(rhyme []string) []string {
    words := len(rhyme)
    proverb := make([]string, words)
    var message string
	for i := 0; i < words; i++ {
        switch {
        case (i + 1) == words:
            message = fmt.Sprintf("And all for the want of a %s.", rhyme[0])
		default: 
            message = fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i], rhyme[i+1])            
        }
        proverb[i] = message
    }
    return proverb
}

File: solutions/csharp/proverb/1/Proverb.cs
public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        var lines = new List<string>();
        for (int i = 0; i < subjects.Length - 1; i++)
        {
            lines.Add(ForWantOf(subjects[i], subjects[i + 1]));
        }
        if (subjects.Length > 0)
            lines.Add(AllForWantOf(subjects[0]));

        return [.. lines];
    }

    private static string ForWantOf(string first, string second) => $"For want of a {first} the {second} was lost.";

    private static string AllForWantOf(string first) => $"And all for the want of a {first}.";
}